lib/SIL/SILInstructionWorklist.cpp

```diff
@@ -0,0 +1,38 @@
+//===--- SILInstructionWorklist.cpp ---------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#define DEBUG_TYPE "sil-instruction-worklist"
+#include "swift/SIL/SILInstructionWorklist.h"
+
+using namespace swift;
+
+void SILInstructionWorklist::add(SILInstruction *instruction) {
+  if (!worklistMap.insert(std::make_pair(instruction, worklist.size())).second)
+    return;
+
+  LLVM_DEBUG(llvm::dbgs() << loggingName << ": ADD: " << *instruction << '\n');
+  worklist.push_back(instruction);
+}
+
+void SILInstructionWorklist::addInitialGroup(ArrayRef<SILInstruction *> list) {
+  assert(worklist.empty() && "worklist must be empty to add initial group");
+  worklist.reserve(list.size() + 16);
+  worklistMap.reserve(list.size());
+  LLVM_DEBUG(llvm::dbgs() << loggingName << ": ADDING: " << list.size()
+                          << " instrs to worklist\n");
+  while (!list.empty()) {
+    SILInstruction *instruction = list.back();
+    list = list.slice(0, list.size() - 1);
```

Comment:
I am fine if you want to fix this in a subsequent commit, but we should modernize this code a little. Specifically:

1. Instead of using ArrayRef<T>.slice(), we can use a method that specifically does this ArrayRef<T>.drop_last(). See: http://llvm.org/doxygen/classllvm_1_1ArrayRef.html#a1c7ded859212582e33828a47259f9bd3.
2. Rather than using insert with std::make_pair, worklistMap IIRC has a method called try_emplace that will let you avoid making the pair since it forwards the arguments to the underlying pair constructor.

so with both of these changes you would have:

```
list = list.drop_back()
worklistMap.try_emplace(instruction, worklist.size());
```

Reply:
Sounds good.  Will go into the forthcoming cleanup commit.

Reply:
Changed the worklist to use a blot set vector so the second call becomes an insert in https://github.com/apple/swift/pull/27028 .